presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimateMath.java

```diff
@@ -20,81 +20,6 @@
-    }
-
-    public static PlanNodeStatsEstimate differenceInNonRangeStats(PlanNodeStatsEstimate left, PlanNodeStatsEstimate right)
-    {
-        return differenceInStatsWithRangeStrategy(left, right, ((leftRange, rightRange) -> leftRange));
-    }
-
-    private static PlanNodeStatsEstimate differenceInStatsWithRangeStrategy(PlanNodeStatsEstimate left, PlanNodeStatsEstimate right, RangeSubtractionStrategy strategy)
-    {
-        PlanNodeStatsEstimate.Builder statsBuilder = PlanNodeStatsEstimate.builder();
-        double newRowCount = left.getOutputRowCount() - right.getOutputRowCount();
-
-        Stream.concat(left.getSymbolsWithKnownStatistics().stream(), right.getSymbolsWithKnownStatistics().stream())
-                .forEach(symbol -> {
-                    statsBuilder.addSymbolStatistics(
-                            symbol,
-                            subtractColumnStats(
-                                    left.getSymbolStatistics(symbol),
-                                    left.getOutputRowCount(),
-                                    right.getSymbolStatistics(symbol),
-                                    right.getOutputRowCount(),
-                                    newRowCount,
-                                    strategy));
-                });
-
-        return statsBuilder.setOutputRowCount(newRowCount).build();
-    }
-
-    @Deprecated
-    private static SymbolStatsEstimate subtractColumnStats(
-            SymbolStatsEstimate leftStats,
-            double leftRowCount,
-            SymbolStatsEstimate rightStats,
-            double rightRowCount,
-            double newRowCount,
-            RangeSubtractionStrategy strategy)
-    {
-        StatisticRange leftRange = StatisticRange.from(leftStats);
-        StatisticRange rightRange = StatisticRange.from(rightStats);
-
-        double nullsCountLeft = leftStats.getNullsFraction() * leftRowCount;
-        double nullsCountRight = rightStats.getNullsFraction() * rightRowCount;
-        double totalSizeLeft = (leftRowCount - nullsCountLeft) * leftStats.getAverageRowSize();
-        double totalSizeRight = (rightRowCount - nullsCountRight) * rightStats.getAverageRowSize();
-        double newNullsFraction = Math.max(nullsCountLeft - nullsCountRight, 0.0) / newRowCount;
-        double newNonNullsRowCount = newRowCount * (1.0 - newNullsFraction);
-
-        StatisticRange range = leftRange;
-        double newDistinctValuesCount = leftStats.getDistinctValuesCount();
-        double leftValuesPerDistinctValue = leftRowCount * (1.0 - leftStats.getNullsFraction()) / leftStats.getDistinctValuesCount();
-        double rightValuesPerDistinctValue = rightRowCount * (1.0 - rightStats.getNullsFraction()) / rightStats.getDistinctValuesCount();
-        if (leftValuesPerDistinctValue <= rightValuesPerDistinctValue) {
-            // right values cover all left values for corresponding distinct values
```

Comment:
I don't see that comment in new code. Is the logic preserved?

Reply:
Yes, the logic is preserved. I just lost the comment.

Reply:
Actually I don't think its preserved for logical binary expression.

Reply:
It is not preserved for `OR`, as the NDV logic for or is different.

Reply:
For OR expression this condition is always true, assuming that `row_count / (disjunct_1_ndvs + disjunct_2_ndvs)` is always `<= row_count / conjunct_1_2_ndvs`.

That is not always true for `NOT` though. Because although `original_ndv` is always `<=` than `negated_predicate_ndvs`, the original `row_count` is >= `negated_predicate_row_count`

Reply:
I'm trying to remember why, but i didn't add this check to the `OR` estimation as it seemed to be extra.

Reply:
>  it seemed to be extra.

But that i mean that the check will always be `true`.

Reply:
It isn't extra. This came from customer case. Lets have expression: 'E1 or R2` and:
```
R1=rows(E1) , R2=rows(E2), where intersect(R1, R2) = 0
```
Then if `rows(E1 AND E2)=R1` this check is needed to get remove NDVs for which there are no rows.

Similarly if `rows(E1 AND E2)=R1/2 (half of rows for each NDVs)` then this check is needed to preserve NDVs for which all rows were not subtracted.

Reply:
Could you please explain why the range merge must be under this check?

Reply:
My thoughts about estimating NDV's for OR were next

NDV_AND - number of distinct values that are matched by both predicates

NDV_E1 - number of distinct values that are matched by E1
NDV_E2 - number of distinct values that are matched by E2

NDV_E1_ONLY - number of distinct values that are matched only by E1
NDV_E2_ONLY - number of distinct values that are matched only by E2

NDV_E1 = NDV_E1_ONLY + NDV_AND
NDV_E2 = NDV_E2_ONLY + NDV_AND

NDV_E1 + NDV_E2 = NDV_E1_ONLY + NDV_E2_ONLY + 2xNDV_AND

So it looks like we should always subtract `NDV_AND`  to remove the `2x` multiplier.

Same for number of rows.

For `NOT` expression the situation is different. The number of distinct values must be negated, and not subtracted.

If number of input values is N, and number of input values negated is N / 3, than the result should be 2/3 * N, but only if rows per NDV is less in source than in negated result, because of the reasons you have explained.

Reply:
`OR` implementation is in principle: https://en.wikipedia.org/wiki/Inclusion%E2%80%93exclusion_principle.
However, there are many rows per NDV and depending on filter only some rows might be filtered. For instance, if you have filter expression: `the_column_that_interest_me < 5` then for each NDV all or none rows will be filtered out. If you have expression `some_other_column < 5` then only some rows (from `the_column_that_interest_me` column) will be filtered out per NDV.

> For NOT expression the situation is different. The number of distinct values must be "heuristically" negated, and not subtracted.

This is what `subtractSubset(inputStats, process(node.getValue());` would do.

Reply:
> Could you please explain why the range merge must be under this check?

If no NDVs are removed then original (left) range is preserved. If some NDVs are removed we could adjust the range. There is some faulty logic in `StatisticRange#subtract` that tries to do that, but I would simply remove it.

Reply:
Maybe i'm being really slow, but in case the `IF` check is applied - the result can be incorrect. Consider following situation:

```
INITIAL ROWS COUNT = 400
INITIAL NDV = 2

P1 SELECTED ROWS = 300
P2 SELECTED ROWS = 3

P1 SELECTED NDV = 2
P2 SELECTED NDV = 1

OR (SUM) ROWS = 303
OR (SUM) NDV = 3
OR (SUM) ROWS PER NDV = 101

AND (INTERSECTION) ROWS = 3
AND (INTERSECTION NDV) = 1
AND (INTERSECTION) ROWS PER NDV = 3

IF CHECKING:

`OR (SUM) ROWS PER NDV` IS GREATER THAN `AND (INTERSECTION) ROWS PER NDV`
RESULT NDV = `OR (SUM) NDV = 3`, THAT IS GREATER THAN INITIAL NDV, NOT CORRECT

IF NOT CHECKING:

RESULT DV = (OR (SUM) NDV = 3) - (AND (INTERSECTION NDV) = 1) = 2 - CORRECT
```